File: src/fintendo/api/routes/v1/companies.py
from fastapi import APIRouter, HTTPException, Query

from fintendo.api.schemas.company import CompanySearchResult
from fintendo.api.services.company_service import CompanyService


router = APIRouter(
    prefix="/companies",
    tags=["Companies"],
)


company_service = CompanyService()


@router.get(
    "/search",
    response_model=list[CompanySearchResult],
)
def search_companies(
    q: str = Query(
        min_length=1,
        description="Company name or ticker.",
    ),
) -> list[CompanySearchResult]:
    try:
        return company_service.search(q)

    except ValueError as exc:
        raise HTTPException(
            status_code=400,
            detail=str(exc),
        ) from exc

    except RuntimeError as exc:
        raise HTTPException(
            status_code=502,
            detail=str(exc),
        ) from exc

File: frontend/lib/companies.ts
import { apiFetch } from "@/lib/api";
import type { CompanySearchResult } from "@/types/company";

export async function searchCompanies(
  query: string,
): Promise<CompanySearchResult[]> {
  const trimmedQuery = query.trim();

  if (!trimmedQuery) {
    return [];
  }

  return apiFetch<CompanySearchResult[]>("/api/v1/companies/search", {
    params: {
      q: trimmedQuery,
    },
  });
}
